### PR TITLE
Support for excluding JDBC spans from being captured

### PR DESCRIPTION
**Pull request by tobiasstadler**
## What does this PR do?
Adds support to exclude SQLs from being captured.

## Checklist
- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JDBCConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import co.elastic.apm.agent.configuration.converter.ListValueConverter;
+import co.elastic.apm.agent.matcher.WildcardMatcher;
+import co.elastic.apm.agent.matcher.WildcardMatcherValueConverter;
+import org.stagemonitor.configuration.ConfigurationOption;
+import org.stagemonitor.configuration.ConfigurationOptionProvider;
+
+import java.util.Collections;
+import java.util.List;
+
+public class JDBCConfiguration extends ConfigurationOptionProvider {
```
Comment (tobiasstadler) [inline apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JDBCConfiguration.java:30]: Maybe it is better to rename it to `DbConfiguration` and move it to apm-agent-core

**Comment by felixbarny** [comment]
Have you considered dropping these spans in an ingest node processor instead? See also https://www.elastic.co/guide/en/elasticsearch/reference/master/drop-processor.html

The tradeoff is that this doesn't decrease the overhead of capturing, serializing, and sending the span. But if it's not an overwhelming amount of spans that need to be dropped, it seems reasonable and works across all agents. 

Could you describe your use case in a little more detail? Do you want to drop the spans for connection validation queries? In that case, it seems like an ingest node processor would be a good option.

If that's not an option, it might make sense to make the feature more generic so that it works for all spans with a `db.statement`. The tradeoff is that the spans would still need to be collected and would then be dropped after the fact (for example by calling `span.requestDiscarding()` after the span has ended.

**Comment by tobiasstadler** [comment]
Currently it is only about validation queries, which account for about 30% of our db queries. We may want to exclude out queries as well, but we haven't decided yet.

I didn't think of using an ingest pipeline. Thanks for the tip. We will implement that for the time being.

Yes, a more general solution is probably the better approach. But since we are only using JDBC for DB access, I went for a JDBC only solution.

Dropping the spans after they ended might impact all spans, as for every span it has to be checked if `db.statement` exists.

**Comment by tobiasstadler** [comment]
While the ingest pipeline works fine, I still think excluding statements in the agent is valuable. In our case we can exclude many SQL statements that currently need to be send to the APM-Server and Elasticsearch only to be dropped there.  Maybe we can avoid that occasionally the queue of our APM-Server is full (without tuning the server/increasing its resources).

I am fine if this is an experimental/internal option until a more general solution is available (which might need an issue/pr in the apm repo).

**Comment by tobiasstadler** [comment]
@felixbarny I created elastic/apm#590 to move this forward

**Comment by tobiasstadler** [comment]
@felixbarny Would `sqls_excluded_from_jdbc_instrumentation` being an internal configuration option be an option?

**Pull request closed without merging**